Sources/Networking/HTTPClient/HTTPClient.swift

```diff
@@ -582,6 +624,87 @@ private extension HTTPClient {
+
+        let retryBackoffInterval: TimeInterval = calculateRetryBackoffTime(
+            forResponse: httpURLResponse,
+            retryCount: request.retryCount
+        )
+
+        Logger.debug(
+            NetworkStrings.api_request_queued_for_retry(
+                httpMethod: request.method.httpMethod,
+                path: request.path,
+                backoffInterval: retryBackoffInterval
+            )
+        )
+        self.operationDispatcher.dispatchOnWorkerThread(delay: .timeInterval(retryBackoffInterval)) {
+            let retriedRequest = request.retriedRequest()
+            self.state.modify {
+                $0.queuedRequests.insert(retriedRequest, at: 0)
+            }
+
+            self.beginNextRequest()
+        }
+    }
+
+    internal func shouldRetryRequest(withStatusCode statusCode: HTTPStatusCode) -> Bool {
+        return self.retryOptions.retriableStatusCodes.contains(statusCode)
+    }
+
+    internal func calculateRetryBackoffTime(
+        forResponse httpURLResponse: HTTPURLResponse,
+        retryCount: UInt
+    ) -> TimeInterval {
+        // If this is our first retry and there's no eTag header, ensure there's no delay
+        if httpURLResponse.allHeaderFields[ResponseHeader.eTag] == nil && retryCount == 1 {
+            return TimeInterval(0)
+        }
+
+        // Use the retry after value from the backend if present
+        if let retryAfterHeaderValue = httpURLResponse.allHeaderFields[ResponseHeader.retryAfter.rawValue] as? String {
+            if let retryAfterMS = Double(retryAfterHeaderValue) {
+                return TimeInterval(milliseconds: retryAfterMS)
+            }
+        }
+
+        // Otherwise, use a default value
+        return calculateDefaultExponentialBackoffTimeInterval(withRetryCount: retryCount)
+    }
+
+    internal func calculateDefaultExponentialBackoffTimeInterval(withRetryCount retryCount: UInt) -> TimeInterval {
+        // Never wait for an original request
+        guard retryCount > 0 else { return 0 }
+
+        let baseBackoff = self.retryOptions.baseBackoff
+        if baseBackoff < 1 {
+            // Since base is less than 1, we need to calculate the backoff by multiplying the baseBackoff by
+            // 2^(retryCount-1). This will result in:
+            //
+            // retryCount 1: backoff = baseBackoff * 2^0 = 0.25
+            // retryCount 2: backoff = baseBackoff * 2^1 = 0.5
+            // retryCount 3: backoff = baseBackoff * 2^2 = 1.0
+            return min(baseBackoff * pow(2.0, Double(retryCount - 1)), self.retryOptions.maxBackoff)
+        } else {
+            // If baseBackoff is greater than or equal to 1, just use baseBackoff^retryCount:
+            // retryCount 1: backoff = baseBackoff^1
+            // retryCount 2: backoff = baseBackoff^2
+            // ...
+            return min(pow(baseBackoff, Double(retryCount)), self.retryOptions.maxBackoff)
+        }
```

Comment:
we only have 3 possible values in that we only retry up to three times, we might as well just explicitly set the values